### PR TITLE
Mtc 20200310

### PR DESCRIPTION
**Pull request by landonreed**
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [x] The description lists all applicable issues this PR seeks to resolve
- [x] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

re ibi-group/datatools-ui#544.

### CODE REVIEW

File: src/main/java/com/conveyal/datatools/manager/jobs/MergeFeedsType.java [hunk 1]
```diff
@@ -2,5 +2,5 @@
 
 public enum MergeFeedsType {
```
Comment (binh-dam-ibigroup) [inline src/main/java/com/conveyal/datatools/manager/jobs/MergeFeedsType.java]: Add JSDoc for this enum.

File: src/main/java/com/conveyal/datatools/manager/models/Model.java [hunk 2]
```diff
@@ -39,6 +39,8 @@ public Model () {
 
     public String id;
 
+    public String retrieveId() { return this.id; }
```
Comment (binh-dam-ibigroup) [inline src/main/java/com/conveyal/datatools/manager/models/Model.java]: This should be named `getId()`.

Comment (binh-dam-ibigroup) [reply src/main/java/com/conveyal/datatools/manager/models/Model.java]: (and `retrieveNotes()` should be renamed `getNotes()`)

File: src/main/java/com/conveyal/datatools/manager/models/Project.java [hunk 3]
```diff
@@ -67,6 +67,12 @@
     //   the "pinned" deployment are shown and compared to the most recent version in the feed sources.
     public String pinnedDeploymentId;
 
+    /**
+     * Feed source in which to store regionally merged GTFS feeds. If specified, during a regional feed merge all feeds
+     * in the project will be merged except for the feed source defined here.
+     */
+    public String regionalFeedSourceId;
```
Comment (binh-dam-ibigroup) [inline src/main/java/com/conveyal/datatools/manager/models/Project.java]: I am not sure why getters and setters were not set up in this entire class.

File: src/main/java/com/conveyal/datatools/manager/controllers/api/ProjectController.java [hunk 4]
```diff
@@ -219,21 +220,43 @@ private static Project checkProjectPermissions(Request req, Project project, Str
...
                 continue;
             }
-            // modify feed version to use prepended feed id
-            LOG.info("Adding {} feed to merged zip", fs.name);
+            LOG.info("Adding {} feed to merged zip", feedSource.name);
             feedVersions.add(version);
         }
+        // Check that the latest regionally merged feed does not already contain input feed versions.
+        if (project.regionalFeedSourceId != null) {
+            Set<String> versionIds = feedVersions.stream().map(FeedVersion::retrieveId).collect(Collectors.toSet());
```
Comment (binh-dam-ibigroup) [inline src/main/java/com/conveyal/datatools/manager/controllers/api/ProjectController.java]: Change to `FeedVersion::getId` (see comment in Model.java)

File: src/main/java/com/conveyal/datatools/manager/models/FeedVersion.java [hunk 5]
```diff
@@ -50,6 +51,10 @@
     private static final Logger LOG = LoggerFactory.getLogger(FeedVersion.class);
     // FIXME: move this out of FeedVersion (also, it should probably not be public)?
     public static FeedStore feedStore = new FeedStore();
+    /**
+     * Input feed versions used to create a merged version.
+     */
+    public Set<String> inputVersions;
```
Comment (binh-dam-ibigroup) [inline src/main/java/com/conveyal/datatools/manager/models/FeedVersion.java]: Replace public field with getter and setter.

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
Fix some method names, add enum JSDoc.
Should public fields be turned into getters/setters? (should still work with 
jackson)

File: src/main/java/com/conveyal/datatools/manager/models/Model.java (hunk 2 above)
Comment (landonreed) [reply src/main/java/com/conveyal/datatools/manager/models/Model.java]: Getters unfortunately cannot be named `getId` due to issues with the Mongo codecs (same with setters). Perhaps there should be comments on these `retrieve*` methods instead.

Comment (landonreed) [reply src/main/java/com/conveyal/datatools/manager/models/Model.java]: I could just remove the getter and change the lambda syntax.

File: src/main/resources/logback.xml [hunk 6]
```diff
@@ -1,9 +1,9 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <configuration>
     <appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
-        <layout class="ch.qos.logback.classic.PatternLayout">
-            <Pattern>%d{MMM dd HH:mm:ss.SSS} [%thread] %-5level %logger{36}:%L - %msg%n</Pattern>
-        </layout>
+        <encoder class="ch.qos.logback.classic.encoder.PatternLayoutEncoder">
+            <pattern>%highlight(%-5level) %d{yyyy-MM-dd'T'HH:mm:ss.SSS} %yellow([%thread]) %blue(%logger{36}\(%file:%line\)) %msg%n</pattern>
```
Comment (evansiroky) [inline src/main/resources/logback.xml:5]: +1 for this date format

File: src/main/java/com/conveyal/datatools/manager/jobs/MergeFeedsJob.java [hunk 7]
```diff
@@ -132,29 +135,57 @@
...
+                project.regionalFeedSourceId = regionalFeedSource.id;
+                Persistence.projects.replace(project.id, project);
+            }
+        }
+        // Assign regional feed source or simply the first parent feed source found in the feed version list (these
+        // should all belong to the same feed source if the merge is not regional).
+        this.feedSource = mergeType.equals(REGIONAL)
+            ? regionalFeedSource
+            : feedVersions.iterator().next().parentFeedSource();
+        // Merged version will be null if the new version should not be stored.
```
Comment (evansiroky) [inline src/main/java/com/conveyal/datatools/manager/jobs/MergeFeedsJob.java:182]: Can you also write some JavaDoc about how this class variable can be null up in the variable definition?

**Review by evansiroky — APPROVED** [review]

**Pull request merged**